smm/lotus_adapter.go

```diff
@@ -0,0 +1,266 @@
+    "github.com/filecoin-project/lotus/build"
+    "github.com/filecoin-project/lotus/chain/actors"
+    "github.com/filecoin-project/lotus/chain/actors/aerrors"
+    laddress "github.com/filecoin-project/lotus/chain/address"
+    "github.com/filecoin-project/lotus/chain/store"
+    "github.com/filecoin-project/lotus/chain/types"
+    "github.com/ipfs/go-cid"
+)
+
+type lotusAdapter struct {
+    address      laddress.Address
+    fullNode     api.FullNode
+    listeners    map[StorageMiningEvents]struct{}
+    headChanges  <-chan[]*store.HeadChange
+    context      context.Context
+}
+
+func NewNode(ctx context.Context, fullNode api.FullNode, address Address) (Node, error) {
+    adapter := lotusAdapter{
+        fullNode: fullNode,
+        listeners: make(map[StorageMiningEvents]struct{}), // this needs synchronization
+        context: ctx,
+    }
+    var err error
+    adapter.address, err = laddress.NewFromString(string(address))
+    if err != nil {
+        return nil, err
+    }
+    adapter.headChanges, _ = fullNode.ChainNotify(ctx)
+    go adapter.eventHandler()
+    return adapter, nil
+}
+
+// TODO: Implement this for TipSetKey
+func stateChangesFromHeadChanges(changes []*store.HeadChange) (stateChanges []*StateChange, err error) {
+    return
+}
+
+// TODO: Remove this after TipSetKey appears
+func tipset2statekey(ts *types.TipSet) (stateKey *StateKey, err error) {
+    return
+}
+
+// TODO: Remove this after TipSetKey appears
+func statekey2tipset(ctx context.Context, stateKey *StateKey, fullNode api.FullNode) (ts *types.TipSet, err error) {
+    return
+}
+
+func (adapter lotusAdapter) eventHandler() {
+    for {
+        select {
+        case changes := <-adapter.headChanges:
+            adapter.notifyListeners(changes)
+
+        case <-adapter.context.Done():
+            return
+        }
+    }
+}
+
+func (adapter lotusAdapter) notifyListeners(changes []*store.HeadChange) {
+    stateChanges, err := stateChangesFromHeadChanges(changes)
+    if err != nil {
+        // TODO: Log, nothing else to do
+        return
+    }
+
+    for _, stateChange := range stateChanges {
+        for listener, _ := range adapter.listeners {
+            listener.OnChainStateChanged(stateChange)
+        }
+    }
+}
+
+func (adapter lotusAdapter) SubscribeMiner(ctx context.Context, cb StorageMiningEvents) error {
+    adapter.listeners[cb] = struct{}{}
+    return nil
+}
+
+func (adapter lotusAdapter) UnsubscribeMiner(ctx context.Context, cb StorageMiningEvents) error {
+    _, found := adapter.listeners[cb]
+    if found {
+        delete(adapter.listeners, cb)
+    }
+    return nil
+}
+
+func (adapter lotusAdapter) callMinerActorMethod(ctx context.Context, method uint64, payload []byte) (cid cid.Cid, err error) {
+    // TODO: Validate that using the miner/worker address for both To and From is allowed
+    msg := types.Message{
+        To:       adapter.address,
+        From:     adapter.address,
```

Comment:
The To address must be the miner actor address, the From is the worker account address. 

In order for one adapter to serve multiple modules, these addresses must be taken as parameters.